bigtop-deploy/puppet/modules/spark/templates/spark-defaults.conf

```diff
@@ -13,7 +13,15 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+<% if @master_url -%>
 spark.master <%= @master_url %>
+<% else -%>
+<% if scope['deploy::roles'].include? 'spark-on-yarn' -%>
+spark.master yarn
+<% else -%>
+spark.master spark://<%= @master_host %>:<%= @master_port %>
```

Comment:
Yup @evans-ye, I agree the `spark-master|worker` roles implied standalone, yet it was always configured for yarn by default.  Confusing for sure.

I like your idea of keeping it backwards compat and introducing the new `spark-standalone`.  With that change, I imagine the templates patch would change to something like this:

```diff
+<% if @master_url -%>
 spark.master <%= @master_url %>
+<% else -%>
+<% if scope['deploy::roles'].include? 'spark-standalone' -%>
+spark.master spark://<%= @master_host %>:<%= @master_port %>
+<% else -%>
+spark.master yarn
```